### PR TITLE
tree: Fix bugs with NodeInDocumentConstraint

### PR DESCRIPTION
**Pull request by alex-pardes**
## Description

This fixes a bug where a violated NodeExistsConstraint could be marked as valid when rebasing over a change which did not revive the target node.

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 1]
```diff
@@ -739,9 +743,9 @@ export class ModularChangeFamily
 			const rebaseChild = (
 				child: NodeId | undefined,
 				baseChild: NodeId | undefined,
-				stateChange: NodeExistenceState | undefined,
+				_attachState: NodeAttachState | undefined,
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:746]: Can we remove this?

File: packages/dds/tree/src/feature-libraries/index.ts [hunk 2]
```diff
@@ -117,7 +117,7 @@ export {
 	genericFieldKind,
 	type HasFieldChanges,
 	type NodeExistsConstraint,
-	NodeExistenceState,
+	NodeAttachState,
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/index.ts:120]: Doesn't seem like this export is needed.

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts (hunk 1 above)
Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:746]: Yes, but we'd just add it back soon.

File: packages/dds/tree/src/test/shared-tree/editing.spec.ts [hunk 3]
```diff
@@ -2709,25 +2709,35 @@ describe("Editing", () => {
...
-					cursorForJsonableTreeNode({ type: leaf.string.name, value: "bar" }),
+					cursorForJsonableTreeNode({ type: leaf.string.name, value: "A" }),
 				);
 
 				const tree2 = tree.fork();
 
-				// Remove a
+				const fooPath: FieldUpPath = { parent: rootNode, field: brand("foo") };
+
+				// Remove A and modify the field containing the node existence constraint.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/shared-tree/editing.spec.ts:2719]:
```suggestion
				// Modify the field containing the node existence constraint then remove its ancestor
```

File: packages/dds/tree/src/feature-libraries/index.ts (hunk 2 above)
Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/index.ts]: Removed

File: packages/dds/tree/src/test/shared-tree/editing.spec.ts [hunk 4]
```diff
@@ -2739,64 +2749,49 @@ describe("Editing", () => {
 
 			it("sequence field node exists constraint", () => {
 				const tree = makeTreeFromJson([]);
+				const { undoStack, unsubscribe } = createTestUndoRedoStacks(tree.events);
+
 				const bPath = {
 					parent: undefined,
 					parentField: rootFieldKey,
 					parentIndex: 1,
 				};
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/shared-tree/editing.spec.ts:2758]: `bPath` seems like an odd name given we use this path to refer to the string "D". You could just use `rootNode2` instead.

**Review by yann-achard-MS — APPROVED** [review]

File: packages/dds/tree/src/test/shared-tree/editing.spec.ts (hunk 4 above)
Comment (alex-pardes) [reply packages/dds/tree/src/test/shared-tree/editing.spec.ts:2758]: Fixed

**Pull request merged**